src/lib/update-lock.ts

```diff
@@ -0,0 +1,107 @@
+import * as Bluebird from 'bluebird';
+import * as lockFileLib from 'lockfile';
+import * as _ from 'lodash';
+import { fs } from 'mz';
+import * as path from 'path';
+import * as Lock from 'rwlock';
+
+import constants = require('./constants');
+import { ENOENT, UpdatesLockedError } from './errors';
+
+type asyncLockFile = typeof lockFileLib & {
+	unlockAsync(path: string): Bluebird<void>;
+	lockAsync(path: string): Bluebird<void>;
+};
+const lockFile = Bluebird.promisifyAll(lockFileLib) as asyncLockFile;
+export type LockCallback = (
+	appId: number,
+	opts: { force: boolean },
+	fn: () => PromiseLike<void>,
+) => Bluebird<void>;
+
+function baseLockPath(appId: number): string {
+	return path.join('/tmp/balena-supervisor/services', appId.toString());
+}
+
+export function lockPath(appId: number, serviceName: string): string {
+	return path.join(baseLockPath(appId), serviceName);
+}
+
+function lockFilesOnHost(appId: number, serviceName: string): string[] {
+	return ['updates.lock', 'resin-updates.lock'].map(filename =>
+		path.join(constants.rootMountPoint, lockPath(appId, serviceName), filename),
+	);
+}
+
+const locksTaken: { [lockName: string]: boolean } = {};
+
+// Try to clean up any existing locks when the program exits
+process.on('exit', () => {
+	const result = [];
+	for (const lockName of _.keys(locksTaken)) {
```

Comment:
This can be:
```suggestion
	for (const lockName in locksTaken) {
```

Reply:
That's what I started with, but our typescript config enforces:
```
[tslint] for (... in ...) statements must be filtered with an if statement [forin]
```

which seems massively out of scope

Reply:
I presume the idea behind that rule is to do
```ts
if (locksTaken.hasOwnProperty(lockName) {
	...
}
```
to avoid issues with iterating properties on the prototype but it is a little weird to force

Reply:
Yeah agreed, and we're also using the keys in a slightly odd way to begin with, the keys either exist with a value of true, or they don't exist, so we might as well use an array - I would rather tackle this as part of a separate PR though.